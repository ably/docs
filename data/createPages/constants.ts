export const DEFAULT_LANGUAGE = 'default';
const TEXT_LANGUAGE = 'text';
const HYPERTEXT_LANGUAGE = 'html';
const YETANOTHERMARKUP_LANGUAGE = 'yaml';
const JSON_LANGUAGE = 'json';
const SH_LANGUAGE = 'sh';
const BASH_LANGUAGE = 'bash';
const ZSH_LANGUAGE = 'zsh';
const SHELL_LANGUAGES = [SH_LANGUAGE, BASH_LANGUAGE, ZSH_LANGUAGE];
export const IGNORED_LANGUAGES = [
  DEFAULT_LANGUAGE,
  TEXT_LANGUAGE,
  HYPERTEXT_LANGUAGE,
  YETANOTHERMARKUP_LANGUAGE,
  JSON_LANGUAGE,
  ...SHELL_LANGUAGES,
];
export const IGNORED_LANGUAGES_FOR_DISPLAY = [
  TEXT_LANGUAGE,
  HYPERTEXT_LANGUAGE,
  YETANOTHERMARKUP_LANGUAGE,
  JSON_LANGUAGE,
  ...SHELL_LANGUAGES,
];
