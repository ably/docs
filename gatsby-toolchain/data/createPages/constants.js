const DEFAULT_LANGUAGE = 'default';
const TEXT_LANGUAGE = 'text';
const HYPERTEXT_LANGUAGE = 'html';
const YETANOTHERMARKUP_LANGUAGE = 'yaml';
const JSON_LANGUAGE = 'json';
const IGNORED_LANGUAGES = [DEFAULT_LANGUAGE, TEXT_LANGUAGE, HYPERTEXT_LANGUAGE, YETANOTHERMARKUP_LANGUAGE, JSON_LANGUAGE];
const IGNORED_LANGUAGES_FOR_DISPLAY = [TEXT_LANGUAGE, HYPERTEXT_LANGUAGE, YETANOTHERMARKUP_LANGUAGE, JSON_LANGUAGE];

module.exports = {
    DEFAULT_LANGUAGE,
    IGNORED_LANGUAGES,
    IGNORED_LANGUAGES_FOR_DISPLAY
};
