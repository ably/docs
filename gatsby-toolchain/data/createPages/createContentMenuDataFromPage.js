import { isArray } from "lodash";
import HtmlDataTypes from "../types/html";

const TYPES_TO_ADD_TO_CONTENT_MENU = [ HtmlDataTypes.h2, HtmlDataTypes.h3 ];
const TYPES_TO_LEVEL_MAP ={
    [HtmlDataTypes.h2]: 2,
    [HtmlDataTypes.h3]: 3
};

const idFromName = name => name.replace(/([a-z0-9])([A-Z]+)/g,'$1-$2').replace(/\s+/g, '-').toLowerCase();

const getTextFromPage = ({ data = '' }, text = []) => {
    if(isArray(data)) {
        const count = data.length;
        for(let i = 0; i < count; ++i) {
            const dataObject = data[i];
            text = getTextFromPage(dataObject, text);
        }
        return text;
    }
    text.push(data);
    return text;
};

const createContentMenuDataFromPage = (page, contentMenuData = []) => {
    const {
        data = '',
        attribs: {
            id
        } = {
            id: null
        },
        type
    } = page;
    if(isArray(data)) {
        const count = data.length;
        for(let i = 0; i < count; ++i) {
            const dataObject = data[i];
            contentMenuData = createContentMenuDataFromPage(dataObject, contentMenuData);
        }
    }

    if(TYPES_TO_ADD_TO_CONTENT_MENU.includes(type)) {
        const name = getTextFromPage({ data }).join('');
        let newID = id;
        if(!id) {
            newID = idFromName(name);
            page.id = newID;
        }
        const contentMenuItem = {
            name,
            id: newID,
            level: TYPES_TO_LEVEL_MAP[type]
        };
        contentMenuData.push(contentMenuItem);
    }

    return contentMenuData;
};

export { idFromName, getTextFromPage };
export default createContentMenuDataFromPage;