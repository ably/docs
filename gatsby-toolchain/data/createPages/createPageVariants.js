const { isArray } = require("lodash");

const DEFAULT_LANGUAGE = 'default';
const TEXT_LANGUAGE = 'text';
const HYPERTEXT_LANGUAGE = 'html';
const YETANOTHERMARKUP_LANGUAGE = 'yaml';
const JSON_LANGUAGE = 'json';

const IGNORED_LANGUAGES = [DEFAULT_LANGUAGE, TEXT_LANGUAGE, HYPERTEXT_LANGUAGE, YETANOTHERMARKUP_LANGUAGE, JSON_LANGUAGE];
const IGNORED_LANGUAGES_FOR_DISPLAY = [TEXT_LANGUAGE, HYPERTEXT_LANGUAGE, YETANOTHERMARKUP_LANGUAGE, JSON_LANGUAGE];

// Mutation (modifying the languageSet provided) is much easier here
// Should also be safer given the guarantees of a Set()
const addLanguagesToSet = (languageSet, languageToKeep = DEFAULT_LANGUAGE) => (content) => {
    const {
        data = '',
        attribs: {
            lang = languageToKeep
        } = {
            lang: languageToKeep
        }
    } = content;
    if(isArray(data)) {
        if(lang !== languageToKeep) {
            languageSet.add(lang);
        }
        data.forEach(addLanguagesToSet(languageSet, languageToKeep))
    }
}

const createLanguagePageVariants = (createPage, documentTemplate) => (contentOrderedList, slug) => {
    const languageSet = new Set();
    contentOrderedList.forEach(addLanguagesToSet(languageSet));
    
    languageSet.forEach(lang => {
        createPage({
            path: `/documentation/${slug}/language/${lang}`,
            component: documentTemplate,
            context: {
                // The slug is the canonical slug, not the variant path
                slug,
                language: lang,
                languages: Array.from(languageSet),
                contentOrderedList
            }
        });
    });
    return Array.from(languageSet);
}

module.exports = {
    DEFAULT_LANGUAGE,
    IGNORED_LANGUAGES,
    IGNORED_LANGUAGES_FOR_DISPLAY,
    addLanguagesToSet,
    createLanguagePageVariants
}