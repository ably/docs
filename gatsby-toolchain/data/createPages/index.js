const { flattenContentOrderedList, maybeRetrievePartial } = require("../transform");
const path = require("path");

const createPages = async ({ graphql, actions: { createPage } }) => {
    const documentTemplate = path.resolve(`src/templates/document.js`)
    const result = await graphql(`
      query {
        allFileHtml {
          edges {
            node {
              slug
              contentOrderedList {
                data
                type
              }
            }
          }
        }
      }
    `);
    const retrievePartialFromGraphQL = maybeRetrievePartial(graphql);
    await Promise.all(result.data.allFileHtml.edges.map(async edge => {
      const contentOrderedList = flattenContentOrderedList(await Promise.all(edge.node.contentOrderedList.map(retrievePartialFromGraphQL)));
  
      createPage({
        path: `/${edge.node.slug}`,
        component: documentTemplate,
        context: {
          slug: edge.node.slug,
          contentOrderedList
        },
      });
    }));
  }

module.exports ={ createPages };