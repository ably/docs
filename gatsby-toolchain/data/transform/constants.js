const LATEST_ABLY_API_VERSION = 1.2;
const LATEST_ABLY_API_VERSION_STRING = '1.2';

module.exports = {
    LATEST_ABLY_API_VERSION,
    LATEST_ABLY_API_VERSION_STRING
};