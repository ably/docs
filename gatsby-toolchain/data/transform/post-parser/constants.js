// Shared constants between different post-parse operations
/**
 * Any number of opening or closing XML tags, inc. HTML tags, followed by any number of spaces
 */
const TAG_REGEX_STRING = '(?:<[\\w \\/]+>)*[\\s\\r\\n]*';

module.exports = {
    TAG_REGEX_STRING
}