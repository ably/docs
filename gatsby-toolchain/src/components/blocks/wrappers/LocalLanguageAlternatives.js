import React, { useState } from 'react';
import { DEFAULT_LANGUAGE } from "../../../../data/createPages/constants";
import languageLabels from '../../../maps/language';
import MenuItemButton, { SelectedMenuItemButton } from '../../Menu/MenuItem/MenuItemButton';
import Html from '../Html';
import HorizontalMenuWrapper from './HorizontalMenuWrapper';

const LocalLanguageAlternatives = ({ language, languages, data, children }) => {
    const [selected, setSelected] = useState(children);
    const [selectedLanguage, setSelectedLanguage] = useState(languages[0]);

    const onClick = ({target : { value }}) => {
        setSelected(<Html data={data[value]} />)
        setSelectedLanguage(value);
    };

    const label = language === DEFAULT_LANGUAGE ? 
        `Select:` :
        `No ${languageLabels[language]} example exists, select:`;
    const languageItems = languages.map(lang => ({
        Component: lang === selectedLanguage ? SelectedMenuItemButton : MenuItemButton,
        props: {
            onClick,
            value: lang
        },
        content: languageLabels[lang] ?? ''
    }));

    return <HorizontalMenuWrapper label={ label } items={ languageItems }>
        {selected}
    </HorizontalMenuWrapper>;
};

export default LocalLanguageAlternatives;