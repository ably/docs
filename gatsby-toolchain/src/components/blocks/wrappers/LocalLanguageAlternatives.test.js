import React from 'react';
import TestRenderer from 'react-test-renderer';
import { DEFAULT_LANGUAGE } from '../../../../data/createPages/constants';
import LocalLanguageAlternatives from './LocalLanguageAlternatives';

const rootElement = <LocalLanguageAlternatives
data={{
    'javascript': [
        {
            "data": [
                {
                    "data": "This is JavaScript.",
                    "type": "text",
                    "name": "text"
                }
            ],
            "type": "tag",
            "name": "code",
            "attribs": {
                "lang": "javascript"
            }
        }
    ],
    'ruby': [
        {
            "data": [
                {
                    "data": "This is Ruby.",
                    "type": "text",
                    "name": "text"
                }
            ],
            "type": "tag",
            "name": "code",
            "attribs": {
                "lang": "ruby"
            }
        }
    ],
    'csharp': [
        {
            "data": [
                {
                    "data": "This is C#.",
                    "type": "text",
                    "name": "text"
                }
            ],
            "type": "tag",
            "name": "code",
            "attribs": {
                "lang": "csharp"
            }
        }
    ]
}}
language={DEFAULT_LANGUAGE}
languages={['javascript','ruby','csharp']}
>
This is default content.
</LocalLanguageAlternatives>;

const languageAlternativesRenderer = TestRenderer.create(rootElement);

const languageAlternativesInstance = languageAlternativesRenderer.root;

// Source: https://stackoverflow.com/questions/34204975/react-is-there-something-similar-to-node-textcontent
const getNodeText = node => {
    if (['string', 'number'].includes(typeof node)) return node
    if (node instanceof Array) return node.map(getNodeText).join('')
    if (typeof node === 'object' && node) return getNodeText(node.props.children)
  };

describe('LocalLanguageAlternatives displays statically as expected', () => {
    it('Snapshot of LocalLanguageAlternatives is the same', () => {
        expect(languageAlternativesRenderer.toJSON()).toMatchSnapshot();
    });
    it('Contains default content', () => {
        expect(getNodeText(languageAlternativesInstance)).toEqual('This is default content.');
    });
});
