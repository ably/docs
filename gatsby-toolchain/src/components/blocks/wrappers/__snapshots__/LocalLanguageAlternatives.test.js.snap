// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalLanguageAlternatives displays statically as expected Snapshot of LocalLanguageAlternatives is the same 1`] = `
Array [
  <menu
    className="sc-crHmcD sc-egiyK sc-bqiRlB ikxONr dMzaTA jZxzrU"
  >
    <li
      className="c-menu__item c-menu__label"
    >
      Select:
    </li>
    <li
      className="c-menu__item"
    >
      <button
        className="sc-bdvvtL sc-gsDKAQ sc-dkPtRN kNKKjb UQpwH fOuwZR"
        onClick={[Function]}
        value="javascript"
      >
        JS
      </button>
    </li>
    <li
      className="c-menu__item"
    >
      <button
        className="sc-bdvvtL sc-gsDKAQ kNKKjb UQpwH"
        onClick={[Function]}
        value="ruby"
      >
        Ruby
      </button>
    </li>
    <li
      className="c-menu__item"
    >
      <button
        className="sc-bdvvtL sc-gsDKAQ kNKKjb UQpwH"
        onClick={[Function]}
        value="csharp"
      >
        C#
      </button>
    </li>
  </menu>,
  "This is default content.",
]
`;
